_posts/2020-02-05-keycloak-z-angularem-i-keycloakiem.md

```diff
@@ -0,0 +1,427 @@
+
+Konsola administracyjna będzie dostępna pod adresem: [127.0.0.1:8180/auth/admin](http://127.0.0.1:8180/auth/admin/)  
+Zalogujemy się do niej przy użyciu loginu i hasła zdefiniowanego w `KEYCLOAK_USER` oraz `KEYCLOAK_PASSWORD`, w naszym przypadku `admin:admin`.
+
+#### Konfiguracja serwera
+
+Sama konfiguracja serwera Keycloak, to temat na osobny wpis, dlatego wkorzystamy przygotowaną wcześniej konfigurację.
+
+Eksport konfiguracji serwera został umieszczony w repozytorium projektu demo (`realm/realm-export.json`) i zostanie załadowany podczas uruchomienia serwera.
+
+**Skonfigurowane zostały:**
+
+- realm `SpringBootAngular`
+- client `SpringBootAngularClient`
+  - Access Type: `confidential` - podczas logowania musimy przekazać secret skonfigurowany w `SpringBootAngularClient`
+    - dostępnę są jeszcze opcje `bearer-only` oraz `public` - więcej na ten temat w dokumentacji [(link)](https://www.keycloak.org/docs/latest/server_admin/#oidc-clients)
+  - Valid Redirect URIs: `*` - dla ułatwienia bez ograniczeń
+  - Web Origins: `*` - dla ułatwienia bez ograniczeń
+- rola `user_role` - tylko użytkownik z tą rolą będzie mógł się autoryzować w naszych aplikacjach
+- użytkownik `user` z rolą `user_role` (`user:password`)
+- użytkownik `user2` bez roli `user_role` (`user2:password2`)
+
+### Aplikacja backendowa
+
+Aplikacja backendowa składa się z dwóch kontrolerów:
+
+- `MainController` - wystawia zabezpieczone endpointy
+  - `/api/hello` - zwraca string `Hello from the Backend!`
+  - `/api/logout` - wylogowuje nas po stronie aplikacji backendowej
+- `KeycloakController` - wystawia publiczny endpoint
+  - `api/keycloak/config` - udostępnia wspólną konfigurację dla frontendu
+
+### Aplikacja frontendowa
+
+Aplikacja frontendowa składa się z trzech komponentów:
+
+- `PublicComponent` - dostępny dla wszystkich bez logowania
+- `ProtectedComponent` - zabezpieczony przed nieautoryzowanym użytkownikiem
+- `ToolbarComponent` - menu z przyciskami dla naszej wygody
+
+## Uwierzytelnianie na backendzie
+
+### Zależności
+
+Na backendzie wykorzystamy zależności:
+
+- `org.keycloak:keycloak-adapter-core`
+- `org.keycloak:keycloak-spring-boot-2-adapter`
+- `org.keycloak:keycloak-tomcat-adapter`
+
+Moglibyśmy wykorzystać jedną zależność, zawierającą powyższe adaptery:
+
+- `org.keycloak:keycloak-spring-boot-2-starter`
+
+jednak w najnowszej dostępnej wersji `4.0.0.Final` wykorzystuje ona stare wersje adapterów. Jedną ze zmian w nowszych wersjach adapterów, które wykorzystamy, jest poprawiona walidacja tokenów. Jeśli w przyszłości pojawi się nowa wersja tej zależności, z nowymi wersjami adapterów, to nic nie stoi na przeszkodzie żeby ją wykorzystać.
+
+### Konfiguracja
+
+Oprócz konfiguracji portu aplikacji oraz poziomu logowania adapterów (dzięki czemu zobaczymy w logach co się dokładnie dzieje), musimy skonfigurować `keycloak-adapter-core`.  
+Dzięki wykorzystaniu `keycloak-spring-boot-2-adapter` możemy wszystko skonfigurować w `application.yml`:
```

Comment:
Bardzo często używasz "wykorzystywać" a rzadko "używać" (20 do 3). Możnaby to trochę zrównoważyć, bo szczególnie w tym akapicie te powtórzenia są widoczne.